File: decl-py/tests/internal/checker_test.py
"""checker (tests/internal/checks.json): the checker's boundary — codes
anchored to their declaration, and a clean module reporting nothing."""

from __future__ import annotations

from decl.checker import check_module
from decl.parse import parse_source


def test_anchored() -> None:
    def codes(src: str) -> list:
        return check_module(parse_source(src)["decls"])

    bad = codes("type Bad = 10..3\n")
    assert len(bad) == 1 and bad[0]["code"] == "E4011"
    assert "loc" not in bad[0], "a declaration-level finding carries no range"
    assert any(
        d["code"] == "E3003" and (d["loc"]["sl"], d["loc"]["sc"], d["loc"]["ec"]) == (0, 10, 11)
        for d in codes("const x = y\n")
    )
    assert codes("type T = { a: int }\nexport output t: T = { a: 1 }\n") == []

File: decl-rs/tests/internal/checker_test.rs
//! checker: the checker's boundary — codes anchored to their declaration,
//! and a clean module reporting nothing.
use decl_lang::checker::check_module;
use decl_lang::parse::parse_source;

#[test]
fn anchored() {
    let codes = |src: &str| check_module(&parse_source(src).decls, None, None);
    let bad = codes("type Bad = 10..3\n");
    assert_eq!(bad.len(), 1);
    assert_eq!(bad[0].code.as_deref(), Some("E4011"));
    assert!(
        bad[0].loc.is_none(),
        "a declaration-level finding carries no range"
    );
    assert!(codes("const x = y\n")
        .iter()
        .any(|d| d.code.as_deref() == Some("E3003")
            && d.loc.map(|l| (l.sl, l.sc, l.ec)) == Some((0, 10, 11))));
    assert!(codes("type T = { a: int }\nexport output t: T = { a: 1 }\n").is_empty());
}
